Implement a validity function using JSON schemas

MichaelKim20 opened the pull request:
1. I implemented the automatic creation of the interfaces using JSON Schemas.
    - This was solved using the library package `json-schema-to-typescript`
    - Whenever 'npm start' and 'npm test' are executed, they are automatically generated.
2. I Implemented a validity function using JSON schemas and interfaces
    - This was solved using the library package `ajv`.


Related to #55

MichaelKim20 commented on package.json:
```diff
@@ -22,9 +23,11 @@
     "assert": "^2.0.0",
     "body-parser": "^1.19.0",
     "express": "^4.17.1",
+    "json-schema-to-typescript": "^9.1.1",
```
This is a library that creates an interface using json-schema.

MichaelKim20 commented on package.json:
```diff
@@ -22,9 +23,11 @@
   "dependencies": {
     "@types/sodium-native": "^2.3.5",
     "@types/sqlite3": "^3.1.6",
+    "ajv": "^6.12.3",
```
This is a library that creates a validity function using json-schema and interface.

MichaelKim20 reviewed (commented):
I removed the automatically generated file.
You can see the code here.
https://gist.github.com/MichaelKim20/764cc481a5148c9fbf6bb299ad4ce067

TrustHenry commented on src/modules/data/schemas/Transaction.json:
```diff
@@ -0,0 +1,23 @@
+    }
+  },
+  "additionalProperties": false,
+  "required": ["type", "inputs", "outputs"]
```
What happens if there are no inputs like Genesis Block?

MichaelKim20 replied:
That's a good question.
In the absence of the content, it consists of:
```
{
...
    "inputs": [],
...
}
```

TrustHenry replied:
When validation, Does it not get caught on `required` items?

MichaelKim20 replied:
The transactions in the Genesis block have no elements in the inputs.
The transaction is being handled normally.

MichaelKim20 replied:
The test code below is also being processed normally.
``` TypeScript
        assert.ok(Validator.isValidOtherwiseThrow<ITransaction>
        ("Transaction",
        {
            "type": 0,
            "inputs": [],
            "outputs": [
                {
                    "value": "400000000000",
                    "address": "GA3DMXTREDC4AIUTHRFIXCKWKF7BDIXRWM2KLV74OPK2" +
                        "OKDM2VJ235GN"
                }
            ]
        }));
```

MichaelKim20 replied:
The previous test code was also left.

TrustHenry commented on src/modules/data/validator/index.ts:
```diff
@@ -0,0 +1,121 @@
+{
+    /**
+     * The map of validation functions created to reuse
+     * a once created validation function.
```
```suggestion
     * An once created validation function.
```

MichaelKim20 replied:
```
The map of validation functions created to reuse -
an once created validation function.
```
I'll change it like this. Do you like it?

TrustHenry commented on tests/ValidationJSON.test.ts:
```diff
@@ -0,0 +1,102 @@
+            "outputs": []
+        }));
+
+        // When everything is normal
+        assert.ok(Validator.isValidOtherwiseThrow<ITransaction>("Transaction", {
+            "type": 1,
+            "inputs": [],
+            "outputs": [
+                {
+                    "value": "400000000000",
+                    "address": "GA3DMXTREDC4AIUTHRFIXCKWKF7BDIXRWM2KLV74OPK2OKDM2VJ235GN"
+                }
+            ]
+        }));
```
Input value is required on normal case .

MichaelKim20 replied:
``` typescript
    {
            "type": 0,
            "inputs": [
                {
                    "previous": "0x5d7f6a7a30f7ff591c8649f61eb8a35d034824ed5" +
                        "cd252c2c6f10cdbd2236713dc369ef2a44b62ba113814a9d819" +
                        "a276ff61582874c9aee9c98efa2aa1f10d73",
                    "index": 1,
                    "signature": "0x07557ce0845a7ccbba61643b95e310bd3ae06c41" +
                        "fab9e8761ff3b0e5d28a5d625a3b951223c618910b239e7b779" +
                        "c6c671252a78edff4d0f37bdb25982e4f4228"
                }
            ],
            "outputs": [
                {
                    "value": "400000000000",
                    "address": "GA3DMXTREDC4AIUTHRFIXCKWKF7BDIXRWM2KLV74OPK2" +
                        "OKDM2VJ235GN"
                }
            ]
        }
```

TrustHenry reviewed (approved):
Nice! , Just small nits

MichaelKim20 committed:
Define and apply the class `Height` to fit JSON's structure

MichaelKim20 committed:
Remove old validity function and test code that are no longer needed

MichaelKim20 committed:
Add JSON schema to the data that makes up the block

It will be used to create interfaces.
It will be used to validate the data.

MichaelKim20 committed:
Implement a validity function using JSON schemas

1. The interfaces is automatically generated using JSON schemas.
   - This was solved using the library package 'json-schema-to-typescript'.
   - Whenever 'npm start' and 'npm test' are executed, they are automatically generated.
2. Implement a validation function using JSON schemas and interfaces
   - This was solved using library package 'ajv'.
